src/nvim/api/buffer.c

```diff
@@ -636,6 +634,297 @@ ArrayOf(Integer, 2) buffer_get_mark(Buffer buffer, String name, Error *err)
+///
+/// @param buffer The buffer handle
+/// @param namespace a identifier returned previously with namespace_create
+/// @param id The mark's id
+/// @param[out] err Details of an error that may have occurred
+/// @return [id, row, col]
+ArrayOf(Integer, 3) buffer_mark_next(Buffer buffer, Integer namespace, Object id, Error *err)
+{
+  Array rv = ARRAY_DICT_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  ExtendedMark *extmark = extmark_from_id_or_pos(buffer, namespace, id, err);
+  if (!extmark) {
+    return rv;
+  }
+
+  bool match = 0;
+  ExtendedMark *next = extmark_next(buf, (uint64_t)namespace, extmark->line->lnum, extmark->col, match);
+  if (!next) {
+    ADD(rv, INTEGER_OBJ(-1));
+    ADD(rv, INTEGER_OBJ(-1));
+    ADD(rv, INTEGER_OBJ(-1));
+  } else {
+    ADD(rv, INTEGER_OBJ((Integer)next->mark_id));
+    ADD(rv, INTEGER_OBJ(next->line->lnum));
+    ADD(rv, INTEGER_OBJ(next->col));
+  }
+  return rv;
+}
+
+/// Returns mark info inbetween and including the bounds
+/// If there are no following marks returns [-1, -1, -1] // TODO
+/// To search untill the end of the buffer do an upper bound of -1 //TODO
+///
+/// @param buffer The buffer handle
+/// @param namespace a identifier returned previously with namespace_create
+/// @param lower [row, col] or mark_id of lower bound
+/// @param uower [row, col] or mark_id of upper bound
+/// @param[out] err Details of an error that may have occurred
+/// @return [[mark_id, row, col], ...]
+ArrayOf(Object) buffer_mark_nextrange(Buffer buffer, Integer namespace, Object lower, Object upper, Error *err)
+{
+  Array rv = ARRAY_DICT_INIT;
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  ExtendedMark *l_extmark = extmark_from_id_or_pos(buffer, namespace, lower, err);
+  ExtendedMark *u_extmark = extmark_from_id_or_pos(buffer, namespace, upper, err);
+  if(!l_extmark || !u_extmark) {
+    return rv;
+  }
+  ExtmarkArray *extmarks_in_range = extmark_nextrange(buf, (uint64_t)namespace,
+                                           l_extmark->line->lnum, l_extmark->col,
+                                           u_extmark->line->lnum, u_extmark->col);
+
+  if (!extmarks_in_range) {
+    ADD(rv, INTEGER_OBJ(-9));
+    return rv;
+  }
+
+  Array mark = ARRAY_DICT_INIT;
+  ExtendedMark *extmark;
+  for (size_t i = 0; i < kv_size(*extmarks_in_range); i++) {
+    mark.size = 0;
+    mark.capacity = 0;
+    mark.items = 0;
+    extmark = kv_A(*extmarks_in_range, i);
+    ADD(mark, INTEGER_OBJ( (Integer)extmark->mark_id) );
+    ADD(mark, INTEGER_OBJ(extmark->line->lnum));
+    ADD(mark, INTEGER_OBJ(extmark->col));
+    ADD(rv, ARRAY_OBJ(mark));
+  }
+  return rv;
+}
+
+/// Setup a new namepsace for holding you marks
+///
+/// @param namespace String name for the namespace
+/// @param[out] err Details of an error that may have occurred
+/// @return integer id to be used with future mark_ calls, or 0 if name exists
+Integer buffer_mark_ns_init(String namespace, Error *err)
```

Comment:
These functions which do not operate on a buffer should live in api/vim.c.

This one would be named nvim_mark_init_ns.